byzcoin/statetrie.go

```diff
@@ -149,13 +155,84 @@ func (t *stagingStateTrie) GetVersion() Version {
 	return readVersion(t)
 }
 
+func (t *stagingStateTrie) LoadConfigFromTrie() (*ChainConfig, error) {
+	return t.loadConfigFromTrie(t)
+}
+
+func (t *stagingStateTrie) LoadDarcFromTrie(id darc.ID) (*darc.Darc, error) {
+	return t.loadDarcFromTrie(t, id)
+}
+
+type trieCache struct {
+	config *ChainConfig
+	darcs  map[string]*darc.Darc
+}
+
+func (tc *trieCache) loadConfigFromTrie(st ReadOnlyStateTrie) (*ChainConfig, error) {
+	if tc.config != nil {
+		return tc.config, nil
+	}
+
+	// Find the genesis-darc ID.
+	val, _, contract, _, err := GetValueContract(st, NewInstanceID(nil).Slice())
+	if err != nil {
+		return nil, xerrors.Errorf("reading trie: %w", err)
+	}
+	if string(contract) != ContractConfigID {
+		return nil, xerrors.New("did not get " + ContractConfigID)
+	}
+
+	tc.config = &ChainConfig{}
+	err = protobuf.DecodeWithConstructors(val, tc.config,
+		network.DefaultConstructors(cothority.Suite))
+	if err != nil {
+		return nil, xerrors.Errorf("decoding config: %v", err)
+	}
+
+	return tc.config, nil
+}
+
+func (tc *trieCache) loadDarcFromTrie(st ReadOnlyStateTrie,
+	id darc.ID) (*darc.Darc, error) {
+	if len(tc.darcs) > 0 {
+		if d := tc.darcs[string(id)]; d != nil {
+			return d, nil
+		}
+	} else {
+		tc.darcs = make(map[string]*darc.Darc)
+	}
+	darcBuf, _, contract, _, err := st.GetValues(id)
+	if err != nil {
+		return nil, xerrors.Errorf("reading trie: %v", err)
+	}
+	config, err := tc.loadConfigFromTrie(st)
+	if err != nil {
+		return nil, xerrors.Errorf("reading trie: %v", err)
+	}
+	var ok bool
+	for _, id := range config.DarcContractIDs {
+		if contract == id {
+			ok = true
+		}
+	}
+	if !ok {
+		return nil, xerrors.New("the contract \"" + contract + "\" is not in the set of DARC contracts")
+	}
+	tc.darcs[string(id)], err = darc.NewFromProtobuf(darcBuf)
```

Comment:
Use a local variable before assigning to map, otherwise, it might get polluted by failing call.